### PR TITLE
Adding support for bulk mode

### PR DESCRIPTION
**Pull request by psquickitjayant**

### CODE REVIEW

**Comment by mostlyjason** [comment]
@indexzero  can you help review this one? it fixes some performance and memory mgmt issues users have had.

**Comment by mostlyjason** [comment]
This PR updates #47

**Comment by indexzero** [comment]
@mostlyjason @psquickitjayant there are still a number of style issues in this code, but TravisCI is not setup so it's not picking them up. I have my normal `winston` maintenance meeting / work session next Friday.

I'll get this merged in then but (honestly) it's going to be a lot of work:

- There are no tests.
- I'm going to have to massage a fair amount (e.g. changes to `common.loggly` would be better implemented as `common.logglyBulk` which is switched in `lib/loggly/client.js`
- Still does not conform to project style.

If you can make some updates addressing this between now and then it would be appreciated.

File: lib/loggly/common.js [hunk 1]
```diff
@@ -109,35 +117,75 @@ common.loggly = function () {
...
     requestOptions.headers.authorization = 'Basic ' + new Buffer(auth.username + ':' + auth.password).toString('base64');
   }
-
-  if (requestBody) {
+  
+  function sendBulkLogs(){
+    if(arrMsg.length == 0){
+      return;
+    }
+    requestOptions.body = arrMsg.toString().replace(/,{/g , "\n{");        // Add New line on the message string
```
Comment (indexzero) [inline lib/loggly/common.js]: This should really be implemented as a `\n` delimited JSON stream which is then piped to the outbound request. See [stream-array](https://www.npmjs.com/package/stream-array).

For example: 

```

var streamify = require('stream-array');
var jsonstream = require('jsonstream');
var stringify = jsonstream.stringify(false);

stringify.pipe(process.stdout);

streamify([
  { ok: true },
  { ok: true },
  { ok: true },
  { ok: true },
  { ok: true }
]).pipe(stringify);
```

Outputs:

```
{"ok":true}
{"ok":true}
{"ok":true}
{"ok":true}
{"ok":true}
```

File: lib/loggly/common.js [hunk 2]
```diff
@@ -109,35 +117,75 @@ common.loggly = function () {
...
   }
-
-  try {
-    request(requestOptions, function (err, res, body) {
-      if (err) {
-        return onError(err);
+  
+  
+  if (requestBody) {
+    if(isBulk == true){
```
Comment (indexzero) [inline lib/loggly/common.js]: `if` spacing here. Please remember to run linters.

File: lib/loggly/common.js [hunk 3]
```diff
@@ -109,35 +117,75 @@ common.loggly = function () {
...
+      if(timerFunction == null){
```
Comment (indexzero) [inline lib/loggly/common.js]: `if` spacing here. Please remember to run linters.

File: lib/loggly/common.js [hunk 4]
```diff
@@ -109,35 +117,75 @@ common.loggly = function () {
...
+        timerFunction = setInterval(function(){
```
Comment (indexzero) [inline lib/loggly/common.js]: `function` spacing here. Please remember to run linters.

File: lib/loggly/common.js [hunk 5]
```diff
@@ -109,35 +117,75 @@ common.loggly = function () {
...
+        timerFunction = setInterval(function(){ 
+          sendBulkLogs(); 
+        },30000);
       }
-
-      var statusCode = res.statusCode.toString();
-      if (Object.keys(failCodes).indexOf(statusCode) !== -1) {
-        return onError((new Error('Loggly Error (' + statusCode + '): ' + failCodes[statusCode])));
+      arrMsg.push(requestBody);
+      if(arrMsg.length == arrSize){
```
Comment (indexzero) [inline lib/loggly/common.js]: `if` spacing here. Please remember to run linters.

File: lib/loggly/common.js [hunk 6]
```diff
@@ -109,35 +117,75 @@ common.loggly = function () {
...
+        sendBulkLogs();
       }
-
-      success(res, body);
-    });
-  }
-  catch (ex) {
-    onError(ex);
+    }
+    else{
```
Comment (indexzero) [inline lib/loggly/common.js]: `else` spacing here. Please remember to run linters.

File: lib/loggly/common.js [hunk 7]
```diff
@@ -109,35 +117,75 @@ common.loggly = function () {
...
+    headers: isBulk ? {} : headers || {}             // Set headers empty for bulk
   };
 
   if (auth) {
     requestOptions.headers.authorization = 'Basic ' + new Buffer(auth.username + ':' + auth.password).toString('base64');
   }
-
-  if (requestBody) {
+  
+  function sendBulkLogs(){
```
Comment (indexzero) [inline lib/loggly/common.js]: `s(){` spacing here. Please remember to run linters.

File: lib/loggly/common.js [hunk 8]
```diff
@@ -109,35 +117,75 @@ common.loggly = function () {
...
+    if(arrMsg.length == 0){
```
Comment (indexzero) [inline lib/loggly/common.js]: `if` spacing here. Please remember to run linters.

File: lib/loggly/common.js (hunk 1 above)
Comment (indexzero) [inline lib/loggly/common.js]: No comments after the end of lines please. Put them on the line before.

**Comment by psquickitjayant** [comment]
@indexzero can you please suggest a linter? we are trying to run common-style, but getting no errors. Do you have a sample command? Thanks

**Comment by indexzero** [comment]
@psquickitjayant what is the exact command you are running? This is what I am running:

```
cd ~/Git/nodejitsu/node-loggly
git checkout gh-48
git remote add psquickitjayant git@github.com:psquickitjayant/node-loggly
git pull psquickitjayant master
./node_modules/.bin/common lib/**/*.js lib/*.js 
```

Then I get this: https://gist.github.com/indexzero/3fdd1f3f05af9149a082. The more important issue here is the structural refactor to the logic you have implemented if you have time: changes to `common.loggly` would be better implemented by creating a separate method `common.logglyBulk` which is then called instead of `common.loggly` in `lib/loggly/client.js` **iff** `bulkMode` is enabled.

**Comment by indexzero** [comment]
@psquickitjayant spent the time setting this repository up with TravisCI using `travis encrypt-file` for the secure Loggly credentials for integration tests. If you `rebase` off of `master` and force push to your `master` you should see the failures should up here.

**Comment by psquickitjayant** [comment]
thanks @indexzero  I am running `node pretest` command to run the same script and getting the following errors

```
Running jshint with options: -c C:\Users\Jayant\Documents\GitHub\node-loggly\node_modules\common-style\dist\.jshintrc lib/loggly/client.js lib/loggly/common.js lib/loggly/search.js lib/loggly.js

events.js:72
        throw er; // Unhandled 'error' event
              ^
Error: spawn ENOENT
    at errnoException (child_process.js:1001:11)
    at Process.ChildProcess._handle.onexit (child_process.js:792:34)
```

**Comment by indexzero** [comment]
Not sure what's going on. Likely a problem with your local setup. You can try to fix it by:

1. Starting from a fresh `git` clone
2. `rm -rf node_modules && npm i`

Please include the **exact** commands you are running to clone the environment and gist the output to:

```
git branch
git --no-pager log --max-count=10
npm ls --depth=1
DEBUG="fashion-show*" npm run pretest
```

**Comment by indexzero** [comment]
@psquickitjayant just to be clear it is absolutely something specific to your environment and not a problem with `common` or `node-loggly`. We know this because the tests for `master` are working on TravisCI: https://travis-ci.org/winstonjs/node-loggly

**Comment by psquickitjayant** [comment]
thanks @indexzero , I am resetting the environment and test it

**Comment by psquickitjayant** [comment]
Closing this as we will update the test cases and send you a new pull request.

**Pull request closed without merging**